accepted/0000-handling-peer-conflicts.md

```diff
@@ -0,0 +1,337 @@
+v6.  The `--legacy-peer-deps` flag is _never_ set for `npm ls`, so that npm
+always warns on incorrect package trees when inspecting.
+
+A `peer set` is a set of packages where each package has a peer dependency
+on one of the others.  In other words, it is the recursive resolution of
+`peerDependencies`, given a particular version of a package.  (The "null"
+trivial peer set is just a single package with no `peerDependencies`.)
+
+In a peerDependencies conflict:
+
+- Either the peer set is possible to resolve without conflicts, or it is
+  not.
+- Either the package depending on the peer set has an explicitly stated
+  dependency on a package that conflicts with one of the members of the
+  peer set, or it does not.
+- Either the package depending on the peer set is the root project, or it
+  is a dependency.
+
+In addition to `--legacy-peer-deps`, a user can opt to use the following
+options:
+
+- `--strict-peer-deps` - fail the install on any `peerDependencies`
+  conflict, even if we can use a heuristic to provide a least-surprising
+  resolution.  This flag is `false` by default, meaning that we _will_ use
+  heuristics as described below for dependencies, but not for the root
+  project.
+- `--force` - If `--strict-peer-deps` is not set, then use heuristics to
+  resolve any and all `peerDependencies` conflicts, even those originating
+  from the root project.  When a satisfactory resolution cannot be found,
+  skip the install of the conflicted peer dependency, so that we do install
+  _something_.
+
+### Heuristics
+
+There are three ways a peer dependency conflict can occur.
+
+- The creation of the peer set.
+- The attempt to place the peers in the tree, where a peer conflicts with a
+  non-peer dependency in the tree.
+- The attempt to place the peers in the tree, where a peer conflicts with a
+  peer dependency in the tree.
+
+The default behavior is **Non-strict mode**.
+
+#### Creation of the Peer Set
+
+To create the peer set:
+
+- Create a virtual node with the manifest of the package that depends on
+  the new node to be created.
+- Add the node to the virtual root's children.
+- For each peer dependency the node has, use either the dependency from the
+  node in question, or from the virtual root if one exists, and resolve it
+  to a peer dep node.
+- Add the peer dep node to the virtual root, and continue the process with
+  its unmet peer dependencies as well.
+
+There are two ways that this can result in a conflict.
+
+##### Conflict External to Peer Set
+
+The first: if the virtual root node has a direct dependency on one of the
+nodes in the peer set, which does not satisfy the dependency from within
+the peer set, then we'll end up adding a package to the peer set that does
+not meet the peer dependency of one or more of the other packages in the
+set.
+
+Example:
+
+```
+x -> (y@1, z@2)
+y -> PEER(z@1)
+```
+
+In this example, the `z@1` dependency will be used in non-strict mode if
+`x` is not the root project.
```

Comment:
this sounds a little bit confusing to me, so if `z@1` is picked for the current level it means that `z@2` gets duped under `x` ? (might be worth adding that bit here if that's the case, just for the sake of clarity for future readers)

Reply:
Action: update to show tree in both scenarios.